Enforce that named_tensor_meta_ is non-null only if there is a non-wildcard name

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38808 Remove device guard codegen on TypeDefault.
* #38806 Don't generate DeviceGuard for CPU wrapping code.
* #38739 Remove supports_named_tensor from codegen entirely.
* **#38725 Enforce that named_tensor_meta_ is non-null only if there is a non-wildcard name**

Today, there are two equivalent representations: named_tensor_meta_ is
null, or named_tensor_meta_ is non-null but all of the dimension names
are wildcard.  Let's reduce the opportunity for behavior divergence by
making the second representation illegal.

This will make it easier for me to add a dispatch key for named
tensor as I can rely on setters to always go through TensorImpl to
maintain invariants on DispatchKey.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21662641](https://our.internmc.facebook.com/intern/diff/D21662641)

ezyang committed:
Enforce that named_tensor_meta_ is non-null only if there is a non-wildcard name

Today, there are two equivalent representations: named_tensor_meta_ is
null, or named_tensor_meta_ is non-null but all of the dimension names
are wildcard.  Let's reduce the opportunity for behavior divergence by
making the second representation illegal.

This will make it easier for me to add a dispatch key for named
tensor as I can rely on setters to always go through TensorImpl to
maintain invariants on DispatchKey.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Enforce that named_tensor_meta_ is non-null only if there is a non-wildcard name"


Today, there are two equivalent representations: named_tensor_meta_ is
null, or named_tensor_meta_ is non-null but all of the dimension names
are wildcard.  Let's reduce the opportunity for behavior divergence by
making the second representation illegal.

This will make it easier for me to add a dispatch key for named
tensor as I can rely on setters to always go through TensorImpl to
maintain invariants on DispatchKey.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/core/NamedTensor.h:
```diff
@@ -16,38 +16,55 @@ namespace at {
+// non-wildcard
+struct CAFFE2_API NamedTensorMeta final : public c10::NamedTensorMetaInterface {
+  // This enum is to remind people that the invariant on constructors is that
+  // the list of dimnames must have at least one non-wilddcard
```
nit: wilddcard -> wildcard

zou3519 commented on aten/src/ATen/core/NamedTensor.h:
```diff
@@ -16,38 +16,55 @@ namespace at {
 
-  void set_names(DimnameList new_names) {
+  void check_invariants() const {
+    // TORCH_INTERNAL_ASSERT_DEBUG_ONLY(
```
Maybe delete this comment?

zou3519 reviewed (approved):
lgtm, thanks!

ezyang committed:
Update on "Enforce that named_tensor_meta_ is non-null only if there is a non-wildcard name"


Today, there are two equivalent representations: named_tensor_meta_ is
null, or named_tensor_meta_ is non-null but all of the dimension names
are wildcard.  Let's reduce the opportunity for behavior divergence by
making the second representation illegal.

This will make it easier for me to add a dispatch key for named
tensor as I can rely on setters to always go through TensorImpl to
maintain invariants on DispatchKey.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]
